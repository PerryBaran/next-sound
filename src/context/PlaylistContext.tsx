'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
