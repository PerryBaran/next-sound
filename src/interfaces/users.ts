interface Songs {
  name: string
  url: string
  id: string
  createdAt: string
}

interface Albums {
  id: string
  name: string
  url: string
  UserId: string
  Songs: Songs[]
  createdAt: string
}

export default interface Users {
  id: string
  name: string
  email: string
  Albums: Albums[]
  createdAt: string
}
