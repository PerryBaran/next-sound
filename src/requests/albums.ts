import {
  postRequest,
  getRequest,
  patchRequest,
  deleteRequest,
  getByIdRequest
} from "./helpers"

export async function postAlbums(data: { name: string; image?: File }) {
  const response = await postRequest("albums", data)
  return response
}

export async function getAlbums(query?: {
  name?: string
  exact?: boolean
  limit?: number
}) {
  const response = await getRequest("albums", query)
  return response
}

export async function getAlbumById(id: string) {
  const response = await getByIdRequest("albums", id)
  return response
}

export async function patchAlbum(
  id: string,
  data: { name?: string; image?: File }
) {
  const response = await patchRequest("albums", id, data)
  return response
}

export async function deleteAlbum(id: string) {
  const response = await deleteRequest("albums", id)
  return response
}
