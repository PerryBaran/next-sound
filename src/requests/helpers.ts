import Cookie from 'js-cookie'
import axios from 'axios'

const instance = axios.create({
  withCredentials: true,
  baseURL: 'https://soundcloud-clone-api.onrender.com'
})

function config() {
  return {
    headers: {
      userToken: Cookie.get('userToken')
    }
  }
}

type Model = 'users' | 'users/login' | 'users/signup' | 'albums' | 'string'

function createForm(model: string, data: any) {
  const formData = new FormData()
  switch (model) {
    case 'songs': {
      if (data.name) {
        formData.append('name', data.name)
      }
      if (data.audio) {
        formData.append('audio', data.audio)
      }
      if (data.position !== undefined) {
        formData.append('position', data.position)
      }
      if (data.AlbumId) {
        formData.append('AlbumId', data.AlbumId)
      }
      break
    }
    case 'albums': {
      if (data.name) {
        formData.append('name', data.name)
      }
      if (data.image) {
        formData.append('image', data.image)
      }
      break
    }
    default: {
      return data
    }
  }
  return formData
}

function error(err: any | { response: { data: { message: string } } }) {
  if (err.response?.data?.message) {
    throw new Error(err.response.data.message)
  }
  throw new Error('Unexpected Error')
}

export async function postRequest(model: Model, data: any) {
  const formData = createForm(model, data)

  try {
    const response = await instance.post(`/${model}`, formData, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function getRequest(
  model: Model,
  query:
    | {
        name?: string
        exact?: boolean
        limit?: number | string
      }
    | undefined
) {
  let endpoint = `/${model}`
  const queryArray = []
  if (query) {
    if (query.name) {
      queryArray.push(`name=${query.name}`)
      if (query.exact) {
        queryArray.push('exact=true')
      }
    }
    if (query.limit) {
      queryArray.push(`limit=${query.limit}`)
    }
    if (queryArray.length > 0) {
      const query = queryArray.join('&')
      endpoint = `${endpoint}?${query}`
    }
  }

  try {
    const response = await instance.get(endpoint, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function getByIdRequest(model: Model, id: string | number) {
  try {
    const response = await instance.get(`/${model}/${id}`, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function patchRequest(
  model: Model,
  id: string | number,
  data: any
) {
  const formData = createForm(model, data)

  try {
    const response = await instance.patch(`/${model}/${id}`, formData, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function deleteRequest(model: Model, id: string | number) {
  try {
    const response = await instance.delete(`/${model}/${id}`, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}
