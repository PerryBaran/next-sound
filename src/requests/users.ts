import {
  postRequest,
  getRequest,
  getByIdRequest,
  patchRequest,
  deleteRequest
} from "./helpers"
import Cookies from "js-cookie"

export async function login(data: { email: string; password: string }) {
  const response = await postRequest("users/login", data)
  Cookies.set("userToken", response.userToken, { expires: 5 })
  return response
}

export async function signup(data: {
  name: string
  email: string
  password: string
}) {
  const response = await postRequest("users/signup", data)
  Cookies.set("userToken", response.userToken, { expires: 5 })
  return response
}

export async function getUsers(query?: {
  name?: string
  exact?: boolean
  limit?: number | string
}) {
  const response = await getRequest("users", query)
  return response
}

export async function getUserById(id: string) {
  const response = await getByIdRequest("users", id)
  return response
}

export async function patchUser(
  id: string,
  data: {
    name?: string
    email?: string
    password?: string
  }
) {
  const response = await patchRequest("users", id, data)
  return response
}

export async function deleteUser(id: string, password: string) {
  const response = await deleteRequest("users", `${id}/${password}`)
  return response
}
