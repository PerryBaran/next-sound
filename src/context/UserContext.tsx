'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
