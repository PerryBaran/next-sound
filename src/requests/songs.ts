import { postRequest, getRequest, patchRequest, deleteRequest } from "./helpers/requestHelpers"

export async function postSongs(data: {
  name: string
  audio: File
  position: number
  AlbumId: string
}) {
  const response = await postRequest("songs", data)
  return response
}

export async function getSongs(query?: {
  name?: string
  exact?: boolean
  limit?: number
}) {
  const response = await getRequest("songs", query)
  return response
}

export async function patchSong(
  id: string,
  data: { name?: string; audio?: File; position?: number }
) {
  const response = await patchRequest("songs", id, data)
  return response
}

export async function deleteSong(id: string) {
  const response = await deleteRequest("songs", id)
  return response
}
