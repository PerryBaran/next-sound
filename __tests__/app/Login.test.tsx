import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import { createContext, useContext } from "react"
import Login from "../../src/app/login/page"
import * as userRequests from "../../src/requests/users"
import * as navigation from "next/navigation"
import * as UserContext from "../../src/context/UserContext"

describe("SignUp", () => {
  let loginSpy: jest.SpyInstance
  let mockPush = jest.fn()
  let mockHandleLogin = jest.fn()

  afterEach(() => {
    jest.clearAllMocks()
  })

  beforeEach(() => {
    mockPush = jest.fn()
    jest.spyOn(navigation, "useRouter").mockReturnValue({
      back: jest.fn(),
      forward: jest.fn(),
      push: mockPush,
      refresh: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn()
    })

    loginSpy = jest.spyOn(userRequests, "login")

    mockHandleLogin = jest.fn()
    const mockContext = createContext({
      user: { name: "", id: "" },
      handleLogin: mockHandleLogin
    })
    jest
      .spyOn(UserContext, "useUserContext")
      .mockImplementation(() => useContext(mockContext))
  })

  describe("no errors", () => {
    const mockUser = {
      name: "name",
      id: "id"
    }

    beforeEach(() => {
      loginSpy.mockResolvedValue(mockUser)

      render(<Login />)
    })

    test("renders correctly", () => {
      expect(screen.getAllByText(/login/i)[0]).toBeInstanceOf(
        HTMLHeadingElement
      )
      expect(screen.getByLabelText(/email/i)).toHaveAttribute("type", "text")
      expect(screen.getByLabelText(/password/i)).toHaveAttribute(
        "type",
        "password"
      )
      expect(screen.getAllByText(/login/i)[1]).toHaveAttribute("type", "submit")
      expect(screen.getByText(/don't have an account\n?/i)).toBeInTheDocument()
      expect(screen.getByText(/signup here/i)).toHaveAttribute(
        "href",
        "/signup"
      )
    })

    test("handleSubmit - no email provided", () => {
      fireEvent.click(screen.getAllByText(/login/i)[1])

      expect(
        screen.getByText(/please provide your email address/i)
      ).toBeInTheDocument()
    })

    test("handleSubmit - no password provided", () => {
      fireEvent.change(screen.getByLabelText(/email/i), {
        target: { value: "email" }
      })
      fireEvent.click(screen.getAllByText(/login/i)[1])

      expect(
        screen.getByText(/please provide your password/i)
      ).toBeInTheDocument()
    })

    test("handleSubmit - invalid email provided", () => {
      fireEvent.change(screen.getByLabelText(/email/i), {
        target: { value: "email" }
      })
      fireEvent.change(screen.getByLabelText(/password/i), {
        target: { value: "12345678" }
      })
      fireEvent.click(screen.getAllByText(/login/i)[1])

      expect(
        screen.getByText(/please provide a valid email/i)
      ).toBeInTheDocument()
    })

    test("handleSubmit - successfull", async () => {
      const mockData = {
        email: "email@email.com",
        password: "12345678"
      }
      fireEvent.change(screen.getByLabelText(/email/i), {
        target: { value: mockData.email }
      })
      fireEvent.change(screen.getByLabelText(/password/i), {
        target: { value: mockData.password }
      })
      fireEvent.click(screen.getAllByText(/login/i)[1])

      await waitFor(() => {
        expect(loginSpy).toBeCalledWith(mockData)
        expect(mockHandleLogin).toBeCalledWith(mockUser)
        expect(mockPush).toBeCalledWith(`profile/${mockUser.name}`)
      })
    })
  })

  describe("errors", () => {
    beforeEach(() => {
      render(<Login />)
    })

    test("handleSubmit throws expected error", async () => {
      const error = "error"
      loginSpy.mockRejectedValue({ message: error })

      fireEvent.change(screen.getByLabelText(/email/i), {
        target: { value: "email@email.com" }
      })
      fireEvent.change(screen.getByLabelText(/password/i), {
        target: { value: "12345678" }
      })
      fireEvent.click(screen.getAllByText(/login/i)[1])

      await waitFor(() => {
        expect(screen.getByText(error)).toBeInTheDocument()
      })
    })

    test("handleSubmit throws unexpected error", async () => {
      loginSpy.mockRejectedValue("")

      fireEvent.change(screen.getByLabelText(/email/i), {
        target: { value: "email@email.com" }
      })
      fireEvent.change(screen.getByLabelText(/password/i), {
        target: { value: "12345678" }
      })
      fireEvent.click(screen.getAllByText(/login/i)[1])

      await waitFor(() => {
        expect(screen.getByText(/unexpected error/i)).toBeInTheDocument()
      })
    })
  })

  test("snapshot", () => {
    const { asFragment } = render(<Login />)

    expect(asFragment()).toMatchSnapshot()
  })
})
