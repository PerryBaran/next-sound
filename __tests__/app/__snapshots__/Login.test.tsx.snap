// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <h2>
        Login
      </h2>
      <label
        for="email"
      >
        <span>
          Email
        </span>
        <input
          id="email"
          name="email"
          type="text"
          value=""
        />
      </label>
      <label
        for="password"
      >
        <span>
          Password
        </span>
        <input
          id="password"
          name="password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Login
      </button>
    </form>
    <p>
      Don't have an account? 
      <a
        href="/signup"
      >
        Signup here
      </a>
    </p>
  </div>
</DocumentFragment>
`;
